haystack_experimental/components/embedders/cohere_doc_image_embedder.py

```diff
@@ -0,0 +1,147 @@
+        """
+        Deserializes the component from a dictionary.
+
+        :param data:
+            Dictionary to deserialize from.
+        :returns:
+                Deserialized component.
+        """
+        init_params = data.get("init_parameters", {})
+        deserialize_secrets_inplace(init_params, ["api_key"])
+
+        # Convert embedding_type string to EmbeddingTypes enum value
+        init_params["embedding_type"] = EmbeddingTypes.from_str(init_params["embedding_type"])
+
+        return default_from_dict(cls, data)
+
+    @component.output_types(documents=List[Document])
+    def run(self, documents: List[Document]):
+        """
+        Embed a list of `Documents`.
+
+        :param documents: documents to embed.
+        :returns:  A dictionary with the following keys:
+            - `documents`: documents with the `embedding` field set.
+        :raises TypeError: if the input is not a list of `Documents`.
+        """
+        if not isinstance(documents, list) or (documents and not isinstance(documents[0], Document)):
+            msg = (
+                "CohereDocumentEmbedder expects a list of Documents as input."
+                "In case you want to embed a string, please use the CohereTextEmbedder."
+            )
+            raise TypeError(msg)
+
+        if not documents:
+            # return early if we were passed an empty list
+            return {"documents": []}
+
+        images_to_embed = []
+        for doc in documents:
+            image_path = doc.meta.get(self.meta_field_for_image_path)
+            if not image_path:
+                raise ValueError(f"Image path not found in document metadata for document {doc.id}")
+            image_content = ImageContent.from_file_path(file_path=image_path)
+            images_to_embed.append(f"data:{image_content.mime_type};base64,{image_content.base64_image}")
+
+        all_embeddings = []
+        for img in images_to_embed:
+            response = self._client.embed(
+                texts=None,
+                images=[img],  # Cohere allows only one image per request
+                model=self.model,
+                input_type="image",
+                truncate="NONE",
+                embedding_types=[self.embedding_type.value],
+            )
+            ## response.embeddings always returns 5 tuples, one tuple per embedding type
+            ## let's take first non None tuple as that's the one we want
+            for emb_tuple in response.embeddings:
+                # emb_tuple[0] is a str denoting the embedding type (e.g. "float", "int8", etc.)
+                if emb_tuple[1] is not None:
+                    # ok we have embeddings for this type, let's take all the embeddings (a list of embeddings)
+                    all_embeddings.extend(emb_tuple[1])
+
+        for doc, embedding in zip(documents, all_embeddings):
+            doc.embedding = embedding
+            doc.meta["embedding_source"] = f"meta.{self.meta_field_for_image_path}"
```

Comment:
Out of curiosity why add this `embedding_source` field?

Reply:
ahh okay I see. I think we might want to make it called `embedding_type` and give it values like `image` vs. `text`. I think `embedding_source` just point to the file path doesn't given enough info to know what was used to make the embedding.

Reply:
Good idea.